Sources/Workspace/DefaultPluginScriptRunner.swift

```diff
@@ -178,52 +181,86 @@ public struct DefaultPluginScriptRunner: PluginScriptRunner {
+        // Construct the command line.  We just pass along any arguments intended for the plugin.
+        var command = [compiledExec.pathString] + pluginArguments
         command += [String(decoding: input, as: UTF8.self)]
 
         // If enabled, run command in a sandbox.
         // This provides some safety against arbitrary code execution when invoking the plugin.
         // We only allow the permissions which are absolutely necessary.
         if self.enableSandbox {
-            command = Sandbox.apply(command: command, writableDirectories: writableDirectories)
+            command = Sandbox.apply(command: command, writableDirectories: writableDirectories + [self.cacheDir])
         }
 
-        // Invoke the plugin script as a subprocess.
-        let result: ProcessResult
+        // Create and configure a Process.
+        let process = Process()
+        process.launchPath = command.first!
+        process.arguments = Array(command.dropFirst())
+        process.environment = ProcessInfo.processInfo.environment
+        process.currentDirectoryURL = self.cacheDir.asURL
+            
+        // Create a dispatch group for waiting until the process has terminated and the data has been read.
+        let waiters = DispatchGroup()
+        // Set up for capturing stdout data.
+        let stdoutPipe = Pipe()
+        var stdoutData = Data()
+        waiters.enter()
+        stdoutPipe.fileHandleForReading.readabilityHandler = { (fileHandle: FileHandle) -> Void in
+            let newData = fileHandle.availableData
+            if newData.isEmpty {
+                fileHandle.readabilityHandler = nil
+                waiters.leave()
+            }
+            else {
+                stdoutData.append(contentsOf: newData)
+            }
+        }
+        process.standardOutput = stdoutPipe
+        
+        // Set up for capturing stderr data.
+        waiters.enter()
+        let stderrPipe = Pipe()
+        var stderrData = Data()
+        stderrPipe.fileHandleForReading.readabilityHandler = { (fileHandle: FileHandle) -> Void in
+            let newData = fileHandle.availableData
+            if newData.isEmpty {
+                fileHandle.readabilityHandler = nil
+                waiters.leave()
+            }
+            else {
+                stderrData.append(contentsOf: newData)
+            }
+        }
+        process.standardError = stderrPipe
+        
+        // Set up a termination handler.
+        process.terminationHandler = { _ in
+            waiters.leave()
+        }
+
+        // Start the process.
+        waiters.enter()
         do {
-            result = try Process.popen(arguments: command)
+            try process.run()
         } catch {
             throw DefaultPluginScriptRunnerError.subprocessDidNotStart("\(error)", command: command)
         }
 
-        // Collect the output. The `PackagePlugin` runtime library writes the output as a zero byte followed by
```

Comment:
Getting rid of the zero byte and separating the streams into result JSON vs freeform output is more robust and less magic.  We could also go the route of the manifest and pass some file handle in an argument, but that seems a bit redundant with the output streams we already have.  We can consider doing the same with the manifest as we do here to remove the number of moving parts.

Reply:
Note that the plugin arranges for both `stdout` and `stderr` to go to `stderr` after squirreling away the original `stdout` for later writing JSON to.  This works because the PackagePlugin code gets access before calling the plugin.

Reply:
iiuc, this code is an alternative approach to the file writing we do today in the manifest evaluation, correct? I suspect this may be related to the concurrency issue we ran into

Reply:
Yes, that's right — for plugins we can't use the file-writing approach because a plugin needs to be able to call back to the plugin host and then get information back to it again.

The issue with the hang turns out to be https://bugs.swift.org/browse/SR-12080.  On non-Darwin platforms, it isn't apparently a guarantee that the NSFileHandle readability handler gets called with a zero-length data after the pipe is closed.  It seems that the Process-related code in TSC relies on that as well, so it's potentially vulnerable to the same thing.  Looking into what to do about it.

Reply:
Closing stdout and putting a small delay masks the problem.  So it seems that on non-Darwin platforms, there's a race between process termination and detecting that stdout/stderr is closed.  The problem is that without relying on the empty data in the `readabilityHandler` there doesn't seem to be a good way to ensure that all data has been read and the process has terminated.

Reply:
I think I'll need to rework the handler to not count on having an empty `availableData` when the handle is closed.  This does make it tricky to make sure that we wait for both process termination and end of data, since there is no guarantee about the order in which those two arrive (indeed in the the past there were bugs because some data would arrive after the termination handler was called, causing loss of the last part of emitted data).

Reply:
But it does seem that TSC might have this same potential problem on non-Darwin platforms (on Darwin we always get the final empty data, but that could just be an implementation detail).

Reply:
Ok, so an update to that is that it only seems to happen if the subprocess closes stdout before it exits — in that case no zero-data invocation happens.  If stdout is kept open until process exit, then there is an extra zero-data on Linux as well.  Somewhat counterintuitive, but that seems to be the pattern, and provides the answer here.

Reply:
does it make sense to cut our loses here and use a file like with do with the manifest evaluation until the underlying bug is fixed?

Reply:
For this first part, that would make sense.  But for having the plugin communicate back with the host (which I am working on as the next PR on top of this one) a separate file is not sufficient.  However, with those changes there is a clear point at which the plugin says that it's done, so there was can avoid the bug that way.